Add `ActiveRecord.reload_attributes`

abaldwin88 opened the pull request:
### Motivation / Background

PR #48241 detects when certain auto-populated columns should be appended to a `RETURNING` clause. However, in certain cases it may be necessary for the application to explicitly specify additional columns. e.g. Database Triggers.

This feature was discussed in #45736 and briefly with Matthew Draper in the Discord channel.

### Detail

This PR adds the ability for the application to specify those attributes with the `reload_attributes` class method.

### Additional information

Additionally a very minor edge case was handled. The application could try to assign a value to a virtual column which essentially results in a no-op. The values persisted in the database would be the same regardless. However, in doing so the value from `RETURNING` wasn't being reflected on the instance attributes after create. See the newly added test and removal of the `_read_attribute` check.

Performing this nil check with `_read_attribute` will also not be compatible when `RETURNING` is implemented on `UPDATE`.  As it is expected that the instance's attributes will be prepopulated with non-nil values. (See related PR here: https://github.com/rails/rails/pull/48628)

---

Paging: @nvasilevski

abaldwin88 committed:
Add `reload_attributes` to Persistence

Add ability for applications to specify a list of attributes that will be reloaded with a SQL `RETURNING` clause when new records are inserted. Long term the plan for databases that do not support RETURNING is to execute an additional SELECT statement.

Also handles a minor edge case if application code attempts to write to a virtual column. The values from `RETURNING` will always be used to update the instance's attributes.

abaldwin88 commented:
Build error looks like a transitory problem with Sauce Labs throttling.

> 09 06 2023 17:22:45.742:ERROR [launcher.sauce]: Can not start microsoftedge 17.17134 (Windows 10)
>   Failed to start Sauce Connect:
>   'create tunnel' request failed with: URL https://saucelabs.com/rest/v1/rubyonrails/tunnels - 400 (Bad Request) Error: HTTP request failed. Server response: 400: Too many active user tunnels: 10 >= 10 (ref: 0872d2db).

nvasilevski reviewed (commented):
Implementation looks correct to me! I believe this is close to what @matthewd had in mind

However, if I'm not mistaken our long-term intention to have this feature to also impact auto-population of columns on `update`. Additionally the same feature can eventually be used to perform `SELECT` query after record creation & update for adapters that do not support `RETURNING` statement  so we need to find another name for the macro that doesn't tie us to specific action or the underlying implementation. 

Ultimately the abstraction should allow applications to say "I'd like `:col1, :col2` attributes to be consistent with database values whenever possible (on update&create) and using whichever solution is available (`RETURNING` or additional `SELECT`)

abaldwin88 commented on activerecord/lib/active_record/model_schema.rb:
```diff
@@ -423,7 +423,11 @@ def columns
 
       def _returning_columns_for_insert # :nodoc:
-        @_returning_columns_for_insert ||= columns.filter_map do |c|
+        _default_returning_columns_for_insert | _returning_on_create
```
I was heavily debating memoizing this. My original thinking was that technically applications could dynamically change `_returning_on_create` after booting so I erred towards safety. That's a pretty big reach though. I've never seen a valid use case for say redefining a `has_many` association after boot for example.

This is on a hot path so let's take the performance improvement 👍 I'll go ahead and add memoization here.

abaldwin88 commented on activerecord/test/cases/persistence_test.rb:
```diff
@@ -45,17 +45,27 @@ def test_populates_non_primary_key_autoincremented_column_for_a_cpk_model
   end if current_adapter?(:PostgreSQLAdapter)
 
+  def test_refreshes_assigned_returning_attributes_on_creation
+    # Attempting to write rand_number_plus_one is essentially a no-op
```
+1 When I first tested this I was a bit surprised it didn't raise an exception.

abaldwin88 replied:
@nvasilevski FYI looks like that behavior was an intentional concession...

https://github.com/rails/rails/pull/38085#issuecomment-611415580

abaldwin88 commented on activerecord/lib/active_record/persistence.rb:
```diff
@@ -1246,7 +1272,7 @@ def _create_record(attribute_names = self.attribute_names)
 
       returning_columns.zip(returning_values).each do |column, value|
-        _write_attribute(column, value) if !_read_attribute(column)
+        _write_attribute(column, value)
```
Thanks for the double check and explanation. I wasn't entirely sure whether I had missed something.

abaldwin88 commented on activerecord/lib/active_record/persistence.rb:
```diff
@@ -561,6 +565,28 @@ def delete(id_or_array)
+      #   returning_on_create :subtotal
+      #
+      #   returning_on_create :subtotal, :grand_total
+      def returning_on_create(*attributes)
```
Moving away from "returning" sounds good to me. I'll ponder some new naming. "refresh" maybe?  "reload" I strongly associate with `SELECT` statements after years of calling the `reload` method. 

I was thinking we should have two separate configuration attributes between `create` and `update`.  An application could have a trigger on insert but not on update.  It would be simpler though to just do one list.

The sooner we can let the application developer know there's a compatibility problem the better. Should we add a TODO note here to add `validate_returning_support!` at a later time?

abaldwin88 replied:
After reflecting a bit more it is probably better to start simple and do a single macro. We can always add more options later.

I also warmed up to having "reload" in the name. Introducing new vocabulary is going to add more mental overhead than it helps in this case. Most Rails developers are going to immediately understand the implications of "reload" as far as their models are concerned. The goal of ActiveRecord is to abstract away certain database details. They shouldn't have to worry about `RETURNING` vs `SELECT` unless there's a reason to go digging.

So I'm leaning towards the name... `reload_attributes_on_save`

nvasilevski replied:
>I was thinking we should have two separate configuration attributes between create and update. An application could have a trigger on insert but not on update.

Sorry, I missed your comment and posted similar opinion as a standalone comment. I agree on this, long-term we do need an option to separate `update` and `create`. Ultimately it's up to the core team to make final decision on APIs but generally I believe we aim to design APIs as flexible as possible so we can extend them later without unnecessary deprecations

abaldwin88 replied:
@nvasilevski No worries! I'll probably check in a code change tomorrow.

I'm planning to use [after_commit](https://apidock.com/rails/ActiveRecord/Transactions/ClassMethods/after_commit) as an example for the parameters.  So it will be `reload_attributes :foo, on: [:create, :update]`.  (No destroy obviously)

abaldwin88 commented on activerecord/lib/active_record/persistence.rb:
```diff
@@ -561,6 +565,28 @@ def delete(id_or_array)
+      # Note: Columns that are auto-incrementing, virtual, or have a default function do
+      # not need to be specified with this method. They will use +RETURNING+ by default.
+      #
+      # Note: Using this with a database backend that doesn't support +RETURNING+ will
+      # result in an error from malformed queries.
```
Good catch. I just tested locally and confirmed they are silently ignored. I'll update accordingly.

abaldwin88 commented:
@nvasilevski Thanks for the review. I'd like to tackle those two items you mentioned down the line. (returning on update and SELECT for mysql). In the interest of keeping this PR small and focused though I didn't try to implement them here. 

Keeping those long term plans in mind though I agree the naming should be revised. See my replies within the review.

abaldwin88 commented on activerecord/lib/active_record/persistence.rb:
```diff
@@ -561,6 +565,10 @@ def delete(id_or_array)
         delete_by(primary_key => id_or_array)
       end
 
+      def reload_attributes_on_save(*attributes) # :nodoc:
```
Not sure what the workflow preferences are for the Rails team but my thinking is if this PR looks acceptable we can merge now but keep this method undocumented until the two follow-up PRs are complete (RETURNING on update and SELECT for mysql).   Once those are done we can add documentation to make it a public method.

nvasilevski commented:
There is something I realized after leaving comments about naming. If we ever going to support an additional `SELECT` query using the abstraction we are adding here we need to allow flexible configuration as not every column that needs to be auto-assigned on `create` needs to be loaded after an `update`
Imagine having a trigger that populates `populated_on_insert_by_a_trigger_column` on creation, since we want the attribute to have the value after record creation we would use  our new abstraction like `reload_attributes_on_save :populated_on_insert_by_a_trigger_column` However this will cause an unnecessary `SELECT` query on updates even though we know the column doesn't change its value on updates unless explicitly updated.

Which means long-term we need to plan to accomomdate an API that actually allows for per-action config. Maybe the API should be similar to callbacks and accept `:only` or `:except` arguments. But ultimately we would like to be able to separately configure columns that need to be initialized only on `create`, only on `update` and virtual columns will need to always be returned back on create and update

owst commented:
Hi @abaldwin88, thanks for your work on this! 

Just to add a +1 for this PR - we're definitely keen to use it when merged - we have many tables with trigger-generated columns that currently force us to issue sub-optimal reloads after create. If we were recreating the app/DB today, we'd be able to use stored generated columns and the recently-merged virtual column RETURNING support, which would be better all-round.

Do you think there's anything else that can be done to get this merged in time for the 7.1 release? I'm happy to help if I can...

abaldwin88 commented:
@owst Thanks for the kind words!

The key bottleneck for this PR is a Rails team member finding time to review. As you might imagine they're pretty busy folks with a lot of things competing for their attention.

One thing that could help is if you pulled this branch down locally and tested it against your application. Reporting back any issues or otherwise confirming everything works as expected. I can't guarantee doing so would lead to this PR moving ahead imminently but it would help add proof that this patch is working and solves a real-world problem.

owst commented:
> One thing that could help is if you pulled this branch down locally and tested it against your application. Reporting back any issues or otherwise confirming everything works as expected.

@abaldwin88  I've tried this branch with one of our apps today, and it seems to Just Work, which is great!

abaldwin88 commented:
Rebased onto latest from `main`. Fixed conflicts with #49290. 

When #49346 is merged it may be worth evaluating that the `random_number_plus_one` assertion from `test_assigned_reload_attributes_on_create` passes when using the `SQLite3Adapter`.

Also SQLite largely discourages using `BEFORE` triggers: 
https://www.sqlite.org/lang_createtrigger.html#cautions_on_the_use_of_before_triggers

I need to do a little more reading but I believe that means the ideal SQLite solution would be including the `RETURNING` clause for virtual columns and issuing an additional `SELECT` query for `AFTER` trigger columns.

Technically though applications using SQLite could still try to utilize `BEFORE` triggers. Or applications using PostgreSQL could use AFTER triggers for that matter. So we should consider expanding this API so applications can specify whether the values are fetched in the same query or a second follow-up `SELECT` query.

simi commented on activerecord/test/schema/postgresql_specific_schema.rb:
```diff
@@ -46,6 +47,31 @@
+_SQL
+  rescue ActiveRecord::StatementInvalid => e
+    if e.message.include?('language "plpgsql" does not exist')
+      execute "CREATE LANGUAGE 'plpgsql';"
```
Any reason to not make this into schema at the beginning of PG test setup? This make all tests running before (random set) being in environment without this, and all tests (the rest of the random set) with this.

abaldwin88 replied:
@simi 

It's probably a good idea to keep the database configuration concerns together at the top of this file. However, I am confused what you mean by some tests running without this? Could you provide instructions I can execute to see what you mean?

Correct me if I'm wrong, but all the statements defined here get executed only once before the test suite runs? That is to say 'plpgsql' will exist in the environment for every test that runs with the PostgreSQL adapter defined.

If I comment out the silencer here:

https://github.com/rails/rails/blob/e73837a5e680c7eeeef2303f043e809fa478245f/activerecord/test/support/load_schema_helper.rb#L6-L7

and run the suite, the console output seems to support that?

simi replied:
@abaldwin88 ahh, you're right. Sorry for confusion. Also I see this pattern is already used in this file. Even I'm not sure it is best practice to let DB fail hard, rescue and retry, your implementation just copies patterns used around. :+1:

abaldwin88 replied:
@simi No worries - thanks for reviewing!

owst commented:
Hi @abaldwin88 just checking in again after a couple of months to see if there's anything more that can be done to progress/prioritise this PR? Thanks!

abaldwin88 commented:
@owst There's a Rails Discord channel you could try raising this PR in to see if it gets any traction. If there is feedback I'll be happy to address and then resolve conflicts with the main branch.
